expands allowed set of packages to define constants that can be verified as part of metric static analysis

coffeepac opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

expands allowed set of packages to define constants that can be verified as part of metric static analysis

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

partial fix for #81863

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
constants/variables from k8s.io for STABLE metrics is now supported
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

coffeepac committed:
seems to work, needs tests and a lot of cleanup

coffeepac commented:
@logicalhan first cut at this.  There are many things left to do.  for this PR I want to only include packages from:
- stdlib
- k8s.io/kubernetes/pkg
- k8s.io/kubernetes/staging/src

k8s.io/klog and k8s.io/utils packages are giving me grief, not sure why yet.  

I'm leaving packages managed by go itself for now as I need to get a better understanding of how go picks the correct library off disk at compile time to make sure I'm grabbing the right version.

this PR will fail the static analysis check as I didn't update the golden list of stable metrics.

I'm not entirely sure how to create a unit test for this.  current unit tests are all done by passing in a string as a code file, my first guess is to create a `codefixtures` subpackage and put code there or reference existing "live" code as part of the unit test, which is what seems to be happening already.

coffeepac commented on pkg/kubelet/certificate/kubelet.go:
```diff
@@ -60,7 +60,7 @@ func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg
 			Name:           "server_expiration_renew_errors",
 			Help:           "Counter of certificate renewal errors.",
-			StabilityLevel: compbasemetrics.ALPHA,
+			StabilityLevel: compbasemetrics.STABLE,
```
the final PR will not include this change.  this change is being made to exercise the new functionality

coffeepac commented on test/instrumentation/decode_metric.go:
```diff
@@ -182,7 +183,24 @@ func (c *metricDecoder) decodeOpts(expr ast.Expr) (metric, error) {
 					return m, err
 				}
+			case *ast.SelectorExpr:
+				packageName := fmt.Sprintf("%v", v.X)
```
hacky, there's a better way to do this below in this file.  I will update later.

coffeepac replied:
performing an explicit cast instead of SprintF for fame

logicalhan commented:
oh this is kinda awesome, have you tried changing the logic so that we try to parse all metrics instead of just the stable ones?

logicalhan commented:
i'm quite curious how far it gets

logicalhan commented:
/assign

logicalhan commented on test/instrumentation/decode_metric.go:
```diff
@@ -182,7 +183,24 @@ func (c *metricDecoder) decodeOpts(expr ast.Expr) (metric, error) {
+				}
+
 			default:
+				fmt.Fprintf(os.Stdout, "key %s is type %T from Default\n", key, v)
```
i think we have to log this directly into a file, i suspect we rely on stdout to output the stuff into yaml

coffeepac replied:
removed debug line

logicalhan commented on test/instrumentation/main.go:
```diff
@@ -173,3 +179,100 @@ func globalVariableDeclarations(tree *ast.File) map[string]ast.Expr {
+
+	// env gets
+	//GOMODCACHE := os.Getenv("GOMODCACHE")
+	GOROOT := os.Getenv("GOROOT")
+	GOOS := os.Getenv("GOOS")
+	KUBE_ROOT := os.Getenv("KUBE_ROOT")
+	KUBE_URL_ROOT := "k8s.io/kubernetes"
+	KUBE_ROOT_INTERNAL := strings.Replace(KUBE_ROOT, KUBE_URL_ROOT, "", 1) // k8s/k8s refs need this stripped
```
these can be declared in outer scope no?

coffeepac replied:
yeah, make 'em top level `consts` even

coffeepac replied:
moved to top level `const` or `var` as appropriate.  I'm open to suggestions on if there is a better way to pull env variables into this application.  I would normally have a config package or something but I'm new to executing code via `go run` like we are here.

logicalhan commented on test/instrumentation/main.go:
```diff
@@ -173,3 +179,100 @@ func globalVariableDeclarations(tree *ast.File) map[string]ast.Expr {
+			continue
+		} else {
+			// stdlib -> prefix with GOROOT
+			pathPrefix = GOROOT + string(os.PathSeparator) + "src"
```
why not path.join?

coffeepac replied:
will do.

coffeepac replied:
all string concats have been changed to strings.Join

coffeepac commented on test/instrumentation/stability-utils.sh:
```diff
@@ -58,7 +58,7 @@ reset=$(tput sgr0)
   stability_check_setup
   temp_file=$(mktemp)
-  doValidate=$(find_files_to_check | grep -E ".*.go" | grep -v ".*_test.go" | sort | xargs -L 200 go run "test/instrumentation/main.go" "test/instrumentation/decode_metric.go" "test/instrumentation/find_stable_metric.go" "test/instrumentation/error.go" "test/instrumentation/metric.go" -- 1>"${temp_file}")
+  doValidate=$(find_files_to_check | grep -E ".*.go" | grep -v ".*_test.go" | sort | KUBE_ROOT=${KUBE_ROOT} xargs -L 200 go run "test/instrumentation/main.go" "test/instrumentation/decode_metric.go" "test/instrumentation/find_stable_metric.go" "test/instrumentation/error.go" "test/instrumentation/metric.go" -- 1>"${temp_file}")
```
this took me longer to sort out that I would like to admit.

coffeepac committed:
review comments

coffeepac committed:
missed a paren

coffeepac committed:
fix existing unit tests

coffeepac committed:
add happy path tests for two types of imports

coffeepac commented on test/instrumentation/main.go:
```diff
@@ -35,6 +36,17 @@ const (
+	GOOS string = os.Getenv("GOOS")
+	KUBE_ROOT string = os.Getenv("KUBE_ROOT")
+
+	kubeRootDeSuffixed string = kubeRootDesuffix(KUBE_ROOT)
```
converted to function so unit tests can optionally override the value of KUBE_ROOT and use the same package load process to set this value.

coffeepac commented on test/instrumentation/error.go:
```diff
@@ -23,17 +23,19 @@ import (
+	errStableSummary                = "Stable summary metric is not supported"
+	errInvalidNewMetricCall         = "Invalid new metric call, please ensure code compiles"
+	errNonStringAttribute           = "Non string attribute is not supported"
+	errBadVariableAttribute         = "Metric attribute was not correctly set. Please use only global consts in same file"
+	errBadImportedVariableAttribute = "Metric attribute was not correctly set. Please use only global consts in correclty impoprted same file"
```
I _think_ when we can handle all of the different types of importable files, these two can go away as they would be compile time errors (ie:  the symbol is truly unresolveable)

coffeepac commented:
I've added unit tests for a stdlib import (`net/http`) and a k8s import (the original example).  They both work, though I need to fix the second unit test because my home directory only exists on my home.

for this PR i need to add unit tests for the newly created functions, make sure they're sane.  I also need to figure out why klog and utils are not being found.  This seems pretty important.

I haven't tried running against ALPHA level metrics yet.  can do that later, should be fun though, see what else pops out as ready to go.

coffeepac committed:
tests for path resolver, add KUBE_ROOT to both top level func calls

coffeepac committed:
testing patches.  add k8s.io/staging, remove local home

coffeepac commented:
added unit test for local disk import path resolver
added unit test for k8s.io/<not kubernetes> that resolves to k8s.io/staging paths

ToDo:
- klog import resolution
- run against ALPHA metrics

label just changed to XL so this is probably as wide a scope as I should do as part of this PR.

coffeepac committed:
minor cleanups

coffeepac commented:
@logicalhan I'm marking this as RfR.  We might need to wait for the KEP for merge but for now I'd like to make sure these tests all work in CI.

coffeepac commented:
I tried running this against the ALPHA metrics and got a bunch of errors like:
- `Opts for STABLE metric was not directly passed to new metric function` which seems valid
- `Non string attribute is not supported` (these seem to be from a `fmt.Sprintf` function or multiline string)
- Buckets should be set to list of floats, result from function call of prometheus.LinearBuckets or prometheus.ExponentialBucket (for a []float64
- some errInvalidNewMetricCall (I don't fully understand this, it might require a patch to the stability framework)
- a bunch of nil panics because the AST walking is passing a nil object when it shouldn't

and in fact, there were enough nil panics that script didn't finish.  There's probably some work that can be filed out of this.

coffeepac commented:
/hold

this shouldn't be merged until I revert the change to `pkg/kubelet/certificate`, I've only included it to ensure things work correctly in CI.

coffeepac commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

coffeepac commented:
the verify test failed as expected, I haven't added the demo metric to the stable list (and I shouldn't).

verify failed in a bunch of unexpected ways, I'll sort that out soonish.

coffeepac committed:
hack/verify fixes

coffeepac commented:
/test pull-kubernetes-node-e2e-containerd

coffeepac commented:
@logicalhan @brancz this PR adds support to STABLE metrics for including conts/vars from k8s.io (both pkg/ and vendor/) and stdlib.  The failing verify test is expected because that metric isn't mine to control but I've marked it as stable to verify the CI.

If you want I can remove support for stdlib, it was easy and I had already done it so I've left it in for now.  Otherwise, anything else for this PR?

logicalhan commented on test/instrumentation/testdata/staging/src/k8s.io/metrics/metrics.go:
```diff
@@ -0,0 +1,19 @@
+
+package metrics
+
+const OKGO = "ThisIsNotTheSoundOfTheTrain"
```
:)

logicalhan reviewed (commented):
what kind of output do you get if you try to get a full list of metrics that are ALPHA?

ehashman commented:
/cc

coffeepac commented:
@logicalhan ALPHA output was categorized above https://github.com/kubernetes/kubernetes/pull/103654#issuecomment-887959190

coffeepac commented:
/test pull-kubernetes-node-e2e-containerd

logicalhan reviewed (commented):
/lgtm

coffeepac committed:
revert test STABLE declaration

coffeepac commented:
/hold cancel

test STABLE metric label reverted to alpha.

logicalhan reviewed (approved):
/lgtm
/approve

This is great! Thanks for tackling this!

logicalhan commented:
/priority important-soon
/triage accepted

ehashman commented:
/retitle expands allowed set of packages to define constants that can be verified as part of metric static analysis
